Sources/FoundationEssentials/String/StringProtocol+Essentials.swift

```diff
@@ -149,54 +149,96 @@ extension StringProtocol {
     /// paragraph or paragraphs containing a given range.
     @available(macOS 10.10, iOS 8.0, watchOS 2.0, tvOS 9.0, *)
-    public func paragraphRange<R : RangeExpression>(for aRange: R) -> Range<Index> where R.Bound == Index {
-        return String(self).paragraphRange(for: aRange)
+    public func paragraphRange(for range: some RangeExpression<Index>) -> Range<Index> {
+        let r = _paragraphBounds(around: range)
+        return r.start ..< r.end
     }
 }
 
-extension String {
-    internal func lineRange<R : RangeExpression>(for aRange: R) -> Range<Index> where R.Bound == Index {
-
-        // It's possible that passed-in indices are not on unicode scalar boundaries, such as when they're UTF-16 indices.
-        // Expand the bounds to ensure they are so we can meaningfully iterate their UTF8 views.
-        let r = unicodeScalars._boundaryAlignedRange(aRange)
-        let result = utf8._getBlock(for: [.findStart, .findEnd, .stopAtLineSeparators], in: r)
-
-        guard let start = result.start else {
-            guard let end = result.end else {
-                return startIndex ..< endIndex
-            }
-            return startIndex ..< end
-        }
-
-        guard let upper = result.end else {
-            return start ..< endIndex
+extension StringProtocol {
+    @inline(never)
+    internal func _lineBounds(
+        around range: some RangeExpression<Index>
+    ) -> (start: Index, end: Index, contentsEnd: Index) {
+        // Avoid generic paths in the common case by manually specializing on `String` and
+        // `Substring`. Note that we're only ever calling `_lineBounds` on a `Substring`; this is
+        // to reduce the code size overhead of having to specialize it multiple times (at a slight
+        // cost to runtime performance).
+        if let s = _specializingCast(self, to: String.self) {
+            let range = s.unicodeScalars._boundaryAlignedRange(range)
+            return s[...].utf8._lineBounds(around: range)
+        } else if let s = _specializingCast(self, to: Substring.self) {
+            let range = s.unicodeScalars._boundaryAlignedRange(range)
+            return s.utf8._lineBounds(around: range)
+        } else {
+            // Unexpected case. `StringProtocol`'s UTF-8 view is not properly constrained, so we
+            // need to convert `self` to a Substring and carefully convert indices between the two
+            // collections before & after the _lineBounds call.
+            let range = self.unicodeScalars._boundaryAlignedRange(range)
+
+            let startUTF8Offset = self.utf8.distance(from: self.startIndex, to: range.lowerBound)
+            let utf8Count = self.utf8.distance(from: range.lowerBound, to: range.upperBound)
+
+            let s = Substring(self)
+            let start = s.utf8.index(s.startIndex, offsetBy: startUTF8Offset)
+            let end = s.utf8.index(start, offsetBy: utf8Count)
+            let r = s.utf8._lineBounds(around: start ..< end)
```

Comment:
Can you explain more why we need to recalculate `start` and `end`? I thought `range` would be scalar aligned after `_boundaryAlignedRange` on line 185, so it'd also be utf8 aligned, so it'd be safe to use with `s.utf8`. What am I missing?

Reply:
The issue here is that `s` is a different collection from `self`, so the indices we calculated in `s` aren't valid in `self` -- sadly we need to recalculate them from scratch using their UTF-8 offsets.

This whole third case is an exercise in frustration. 🙈 `StringProtocol.UTF8View` isn't guaranteed to be bidirectional (presumably to work around a temporary compiler bug 6-8 years ago, which we forgot to undo before setting things in stone), so we cannot call the generic `_getBounds` on it. And we do not expect or want to ever execute this case, so any and all code size we spend on it is pretty much pure waste. 😖

Reply:
Ah yeah I somehow totally overlooked that we're executing this on this newly constructed `Substring`, not `self`. That makes sense. Thanks!

(This whole third case is sort of unavoidable when we added this function to  `StringProtocol` -- we basically already overpromised what we could do)